[DataPipe] Traverse DataPipe graph excluding primitive and callable

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#67783 [DataPipe] Traverse DataPipe graph excluding primitive and callable**

Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
- This is going to solve a problem that generator objects would also break the `traverse`
  -  Some `IterDataPipe` instances like `Demux` holds a generator object (normally iterator) when It starts iteration. https://github.com/pytorch/pytorch/blob/89b02fc70b6a279b602545fddbcd56b4248cee4f/torch/utils/data/datapipes/iter/combining.py#L220
  - If users start to iterate datapipe first for validation purpose, the generator object is created during iteration. Then, when they send the pipeline into `DataLoader`, the `traverse` is going to break.
- This is also used by `OnDiskCacheHolder` to trace the DataPipe Graph without worry about lambda and generator.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

cc @VitalyFedyunin @ejguan @NivekT

ejguan committed:
[DataPipe] Traverse DataPipe graph excluding primitive and callable

[ghstack-poisoned]

ejguan committed:
Update on "[DataPipe] Traverse DataPipe graph excluding primitive and callable"


Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
This is used by `OnDiskCacheHolder` to trace the DataPipe Graph.

[ghstack-poisoned]

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -225,6 +226,11 @@ def __getattr__(self, attribute_name):
         else:
             raise AttributeError
 
+    def __getstate__(self):
+        if IterableDataset.getstate_hook is not None:
+            return IterableDataset.getstate_hook(self)
+        return self.__dict__
+
```
I really want these APIs moved out of `IterableDataset` by making `IterDataPipe` derived from `IterableDataset`

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -151,6 +151,9 @@ def __len__(self) -> int:
         )
 
     def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+
```
This is needed for all subclasses of `IterDataPipe` which implement `__getstate__`

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

ejguan committed:
Update on "[DataPipe] Traverse DataPipe graph excluding primitive and callable"


Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
This is used by `OnDiskCacheHolder` to trace the DataPipe Graph.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

ejguan committed:
Update on "[DataPipe] Traverse DataPipe graph excluding primitive and callable"


Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
- This is going to solve a problem that generator objects would also break the `traverse`
  -  Some `IterDataPipe` instances like `Demux` holds a generator object (normally iterator) when It starts iteration. https://github.com/pytorch/pytorch/blob/89b02fc70b6a279b602545fddbcd56b4248cee4f/torch/utils/data/datapipes/iter/combining.py#L220
  - If users start to iterate datapipe first for validation purpose, the generator object is created during iteration. Then, when they send the pipeline into `DataLoader`, the `traverse` is going to break.
- This is also used by `OnDiskCacheHolder` to trace the DataPipe Graph without worry about lambda and generator.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

ejguan committed:
Update on "[DataPipe] Traverse DataPipe graph excluding primitive and callable"


Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
- This is going to solve a problem that generator objects would also break the `traverse`
  -  Some `IterDataPipe` instances like `Demux` holds a generator object (normally iterator) when It starts iteration. https://github.com/pytorch/pytorch/blob/89b02fc70b6a279b602545fddbcd56b4248cee4f/torch/utils/data/datapipes/iter/combining.py#L220
  - If users start to iterate datapipe first for validation purpose, the generator object is created during iteration. Then, when they send the pipeline into `DataLoader`, the `traverse` is going to break.
- This is also used by `OnDiskCacheHolder` to trace the DataPipe Graph without worry about lambda and generator.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

cc @VitalyFedyunin @ejguan @NivekT

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).

VitalyFedyunin reviewed (approved):
Looks good, but `set_getstate_hook(None)` little bit counterintuitive, maybe do `disable_getstate_hook`

ejguan committed:
Update on "[DataPipe] Traverse DataPipe graph excluding primitive and callable"


Add `getstate_hook` to exclude primitive objects and callable when serialization when `exclude_primitive` is enabled for `traverse`.
For graph traversing, we don't have to handle the lambda and other stuff.
- This is going to solve a problem that generator objects would also break the `traverse`
  -  Some `IterDataPipe` instances like `Demux` holds a generator object (normally iterator) when It starts iteration. https://github.com/pytorch/pytorch/blob/89b02fc70b6a279b602545fddbcd56b4248cee4f/torch/utils/data/datapipes/iter/combining.py#L220
  - If users start to iterate datapipe first for validation purpose, the generator object is created during iteration. Then, when they send the pipeline into `DataLoader`, the `traverse` is going to break.
- This is also used by `OnDiskCacheHolder` to trace the DataPipe Graph without worry about lambda and generator.

Differential Revision: [D32146697](https://our.internmc.facebook.com/intern/diff/D32146697)

cc @VitalyFedyunin @ejguan @NivekT

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32146697).
